### ISSUE TITLE
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""

### ISSUE THREAD

**Issue by tsteven4** [miurahr/aqtinstall#256]
My application fails with:

> QFactoryLoader::QFactoryLoader() checking directory path "/home/tsteven4/work/pvs2/gui/GPSBabelFE/platforms" ...
> qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
> 
> Aborted (core dumped)

This can be avoided by

1. setting QT_PLUGIN_PATH, e.g. /tmp/Qt/5.12.11/gcc_64/plugins
2. creating a link  /home/qt/work/install -> the installation, e.g. /tmp/Qt/5.12.11/gcc_64

I speculate that this occurs because the qmake binary is not patched.  An official install would have changed these strings in qmake to point to the installation:
qt_prfxpath=/opt/Qt5.12.10/5.12.10/gcc_64
qt_epfxpath=/opt/Qt5.12.10/5.12.10/gcc_64
qt_hpfxpath=/opt/Qt5.12.10/5.12.10/gcc_64

**Comment by tsteven4** [comment]
My speculation is insufficient.  It appears the official installer patches up numerous other strings in other files as well.

It looks like a critical patch is qt_prfxpath in libQt5Core.so.5.12.11

**Comment by miurahr** [comment]
Here is a source place about patching for qmake. An aqt does patch to qmake and QtCore.
 #100 is the issue related. 

What's missing? 

https://github.com/miurahr/aqtinstall/blob/master/aqt/updater.py#L92
https://github.com/miurahr/aqtinstall/blob/master/aqt/updater.py#L127

**Comment by miurahr** [comment]
https://github.com/miurahr/aqtinstall/releases/tag/v1.0.0
aqtinstall v1.0.0 and later has a patch feature.

**Comment by tsteven4** [comment]
Thanks, I seem to  be downlevel and I don't know how to get a more 
recent version.

 From aqtinstall/README.rst at master · miurahr/aqtinstall · GitHub 
<https://github.com/miurahr/aqtinstall/blob/master/README.rst>

> $ pip install -U pip
> $ pip install aqtinstall
got me

> pip show aqtinstall
> Name: aqtinstall
> Version: 0.11.1

**Comment by miurahr** [comment]
```
$ pip install --help

Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...

Install Options:

  -U, --upgrade               Upgrade all specified packages to the newest available version. The handling of dependencies
                              depends on the upgrade-strategy used.

```

**Comment by tsteven4** [comment]
I'm not sure why I had 0.1.11, but I now have 1.1.6.  Despite the log stating

> 2021-06-03 19:38:57,268 - aqt - INFO - Patching /tmp/Qt51211/5.12.11/gcc_64/bin/qmake

> 2021-06-03 19:38:57,402 - aqt - INFO - Patching /tmp/Qt51211/5.12.11/gcc_64/lib/libQt5Core.so.5.12.11


they don't seem to be.
I ran the script test.txt
[test.txt](https://github.com/miurahr/aqtinstall/files/6595194/test.txt)
and the log of std out and std err is testelog.txt
[testlog.txt](https://github.com/miurahr/aqtinstall/files/6595201/testlog.txt)
The .pc files do have the installation path in them, patching those seems to have worked.  But patching the binaries didn't seem to work:

> tsteven4@focal:/tmp$ strings /tmp/Qt51211/5.12.11/gcc_64/lib/libQt5Core.so.5.12.11 | grep qt_prfxpath
qt_prfxpath=/home/qt/work/install
tsteven4@focal:/tmp$ strings /tmp/Qt51211/5.12.11/gcc_64/bin/qmake | grep qt_prfxpath
qt_prfxpath=/home/qt/work/install

**Comment by miurahr** [comment]
You use  Version: 0.11.1

**Comment by tsteven4** [comment]
This looks like a bug to me, I think you meant to use the less than operator instead of greater the than operator.
https://github.com/miurahr/aqtinstall/blob/18fd3da00992acac656030dc96fc48097866cae7/aqt/updater.py#L38-L40

**Comment by tsteven4** [comment]
Using the above suggestion the patch is made, but the file is corrupted.  It seems there is some indexing problem in L42.  Here is a diff of a hexdump of qmake  The patch starts fine but shifts subsequent code.

```
177323,177325c177323,177325
< 002b4aa0  71 74 5f 70 72 66 78 70  61 74 68 3d 2f 68 6f 6d  |qt_prfxpath=/hom|
< 002b4ab0  65 2f 71 74 2f 77 6f 72  6b 2f 69 6e 73 74 61 6c  |e/qt/work/instal|
< 002b4ac0  6c 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |l...............|
---
> 002b4aa0  71 74 5f 70 72 66 78 70  61 74 68 3d 2f 74 6d 70  |qt_prfxpath=/tmp|
> 002b4ab0  2f 51 74 35 31 32 31 31  2f 35 2e 31 32 2e 31 31  |/Qt51211/5.12.11|
> 002b4ac0  2f 67 63 63 5f 36 34 6e  73 74 61 6c 6c 00 00 00  |/gcc_64nstall...|
177341,177343c177341,177343
< 002b4bc0  71 74 5f 65 70 66 78 70  61 74 68 3d 2f 68 6f 6d  |qt_epfxpath=/hom|
< 002b4bd0  65 2f 71 74 2f 77 6f 72  6b 2f 69 6e 73 74 61 6c  |e/qt/work/instal|
< 002b4be0  6c 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |l...............|
---
> 002b4bc0  00 00 00 00 00 00 00 00  00 00 00 00 71 74 5f 65  |............qt_e|
> 002b4bd0  70 66 78 70 61 74 68 3d  2f 68 6f 6d 65 2f 71 74  |pfxpath=/home/qt|
> 002b4be0  2f 77 6f 72 6b 2f 69 6e  73 74 61 6c 6c 00 00 00  |/work/install...|
```

**Comment by tsteven4** [comment]
In addition to the inequality operator change perhaps L42 should be
data = data[:idx] + key + newpath + data[idx + len(key) + len(newpath) :]

**Comment by ddalcino** [comment]
> ```
> data = data[:idx] + key + newpath + data[idx + len(key) + len(newpath) :]
> ```

I'm not too familiar with this code, but I am concerned that there's still a bug here.

Here's a diagram to illustrate:

```
002b4aa0  .....KEY=OLD_VALUE....
                   ^^^^^^^^^
                   └value to be replaced
```

Let's say that `idx` points to the beginning of `KEY`, and `idx + len(key)` points to the first byte of `OLD_VALUE`.

If `newpath` is shorter than `OLD_VALUE`, then you won't overwrite all of `OLD_VALUE`, and you'd have a new value with the tail end of `OLD_VALUE` appended to it. These strings appear to be null-terminated; correct me if I'm wrong, but I imagine that you would need to overwrite the tail end of `OLD_VALUE` with zeros.

This solution is probably fine if you can guarantee that the new value is always longer than the old value, or if qmake ships with an empty space in the `OLD_VALUE` position. From @tsteven4's hexdump diff, it doesn't look like you can make that guarantee though.

**Comment by tsteven4** [comment]
spot on @ddalcino, L42 needs a better fix.  You have to try to get a name that short, but it is doable.  It looks like I fell for fix 2 bugs, add 1.

aqt install --outputdir /t/Q 5.12.11 linux desktop -m qtwebengine
```
 diff qmake.hd2 qmake.hd4
177323,177324c177323,177324
< 002b4aa0  71 74 5f 70 72 66 78 70  61 74 68 3d 2f 68 6f 6d  |qt_prfxpath=/hom|
< 002b4ab0  65 2f 71 74 2f 77 6f 72  6b 2f 69 6e 73 74 61 6c  |e/qt/work/instal|
---
> 002b4aa0  71 74 5f 70 72 66 78 70  61 74 68 3d 2f 74 2f 51  |qt_prfxpath=/t/Q|
> 002b4ab0  2f 35 2e 31 32 2e 31 31  2f 67 63 63 5f 36 34 6c  |/5.12.11/gcc_64l|
```
Feel free to pick it up, I'm tied up for a while.

**Comment by ddalcino** [comment]
I imagine that what you want might look something like this:

```python
begin_val = idx + len(key)
len_writable_space = len(re.match(b"^([^\0]+\0+)[^0]+", data[begin_val:]).group(1))
assert len_writable_space >= 256, "qmake binary has insufficient space for patching"
patched_value = newpath + (len_writable_space - len(newpath)) *  b"\0"
data = data[:begin_val] + patched_value + data[begin_val + len_writable_space:]
```

**Comment by ddalcino** [comment]
> It looks like I fell for fix 2 bugs, add 1.

Story of my life.

Maybe there's some way to write a CI test for this, so we can have actual proof that the fix works? Do you think it would be as simple as:

```bash
$ aqt install --outputdir /t/Q 5.12.11 linux desktop -m qtwebengine
$ # then go build a trivial example program
```

Is there anything I'm missing?

**Comment by miurahr** [comment]
A target value is defined in here [qtbase configure.pri](https://code.qt.io/cgit/qt/qtbase.git/tree/configure.pri?h=5.12#n841)

```
    $${currentConfig}.output.qconfigSource = \
        "/* Installation Info */" \
        "static const char qt_configure_prefix_path_str  [12+256] = \"qt_prfxpath=$$config.input.prefix\";" \
        "$${LITERAL_HASH}ifdef QT_BUILD_QMAKE" \
        "static const char qt_configure_ext_prefix_path_str   [12+256] = \"qt_epfxpath=$$config.input.extprefix\";" \
        "static const char qt_configure_host_prefix_path_str  [12+256] = \"qt_hpfxpath=$$config.input.hostprefix\";" \
        "$${LITERAL_HASH}endif" \
        "" \
```

Is adding  "\0" (null-terminate) enough for fix?

**Comment by miurahr** [comment]
>  CI test 

It is good idea that test target path to be less length.

**Comment by miurahr** [comment]
> Is there anything I'm missing?

Does `qmake -query` help?

```
QT_SYSROOT:
QT_INSTALL_PREFIX:/usr
QT_INSTALL_ARCHDATA:/usr/lib/x86_64-linux-gnu/qt5
QT_INSTALL_DATA:/usr/share/qt5
QT_INSTALL_DOCS:/usr/share/qt5/doc
QT_INSTALL_HEADERS:/usr/include/x86_64-linux-gnu/qt5
QT_INSTALL_LIBS:/usr/lib/x86_64-linux-gnu
QT_INSTALL_LIBEXECS:/usr/lib/x86_64-linux-gnu/qt5/libexec
QT_INSTALL_BINS:/usr/lib/qt5/bin
QT_INSTALL_TESTS:/usr/tests
QT_INSTALL_PLUGINS:/usr/lib/x86_64-linux-gnu/qt5/plugins
QT_INSTALL_IMPORTS:/usr/lib/x86_64-linux-gnu/qt5/imports
QT_INSTALL_QML:/usr/lib/x86_64-linux-gnu/qt5/qml
QT_INSTALL_TRANSLATIONS:/usr/share/qt5/translations
QT_INSTALL_CONFIGURATION:/etc/xdg
QT_INSTALL_EXAMPLES:/usr/lib/x86_64-linux-gnu/qt5/examples
QT_INSTALL_DEMOS:/usr/lib/x86_64-linux-gnu/qt5/examples
QT_HOST_PREFIX:/usr
QT_HOST_DATA:/usr/lib/x86_64-linux-gnu/qt5
QT_HOST_BINS:/usr/lib/qt5/bin
QT_HOST_LIBS:/usr/lib/x86_64-linux-gnu
QMAKE_SPEC:linux-g++
QMAKE_XSPEC:linux-g++
QMAKE_VERSION:3.1
QT_VERSION:5.12.8
```

**Comment by ddalcino** [comment]
> Is adding single b"\0" (null-terminate) enough for fix?

Maybe. It's definitely less cognitive overhead than my solution.

My instinct is to zero out the whole value, but I don't know if that's actually necessary or not. I tend to be overly cautious about modifying binaries.

**Comment by ddalcino** [comment]
Ok, I wrote a test (#259) that does:

```
$ aqt install --outputdir /t/Q 5.12.11 linux desktop
$ # then go build a trivial example program
```

`qt_prfxpath` becomes `/t/Q/5.12.11/gcc_64ll`, which is a non-existent path. Log file: https://dev.azure.com/miurahr/github/_build/results?buildId=4137&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=fcf4a360-d878-547e-2092-53786c55870d&l=262

It builds the example properly, and I don't understand why. It must not have been depending on `qt_prfxpath`.

**Comment by miurahr** [comment]
> It builds the example properly, and I don't understand why. It must not have been depending on qt_prfxpath.

aqtinstall does not only patch qmake and libQt5Core, but also updates `qt.conf` in `bin` directory.
https://github.com/miurahr/aqtinstall/blob/master/aqt/updater.py#L152-L153

qmake will use it.

**Comment by miurahr** [comment]
@tsteven4 might get the issue because BPSBabel project has a [`qt.conf` in the project.](https://github.com/GPSBabel/gpsbabel/blob/695146460d35a5e00efde5b6c2c2cf767181294c/gui/qt.conf)

That has empty prefix then qmake's internal default path may be used, and then got error.

**Comment by tsteven4** [comment]
Regarding a test program this can be used to echo the prefix path
[testbinpatch.zip](https://github.com/miurahr/aqtinstall/files/6604449/testbinpatch.zip)
Note however that if qt.conf is found that will override the built in path. https://doc.qt.io/qt-5/qt-conf.html
Note this likely testing the prefix path the libQt5Core.so, not the prefix path in qmake.

I think the test by @ddalcino may have failed because it didn't use anything that needed to find something using the built in prefix path.  I can't see your test.  The application that I had trouble with is using the xcb plugin on linux.  The prefix path default of the install location is especially useful on the development machine because it allows plugins, translations, etc. to be found without packaging the application.

**Comment by tsteven4** [comment]
> @tsteven4 might get the issue because BPSBabel project has a [`qt.conf` in the project.](https://github.com/GPSBabel/gpsbabel/blob/695146460d35a5e00efde5b6c2c2cf767181294c/gui/qt.conf)
> 
> That has empty prefix then qmake's internal default path may be used, and then got error.

We use that qt.conf when packaging for windows.  It was not influencing the error I reported with the xcb plugin on linux.

**Comment by tsteven4** [comment]
Regarding a test for the qmake prefix patch this can be done by
qmake -query QT_INSTALL_PREFIX
if and only if qt.conf isn't found, e.g. in the same directory as qmake.

**Comment by ddalcino** [comment]
I was trying to use the invalid value stored in `qt_prfxpath` to elicit a build failure in the Azure Pipelines, but this is clearly a test that depends on too many other factors and can easily be foiled.

Both of you are right: A direct query for `QT_INSTALL_PREFIX` more directly tests whether the value was set properly or not. I will switch to this approach.

**Comment by ddalcino** [comment]
Ok, I have more to report regarding this issue:

I was curious as to what effects the `qt_prfxpath`, `qt_epfxpath` and `qt_hpfxpath` have on qmake, and the values reported by `qmake -query`, so I tried modifying the values that @tsteven4's code at #258 patches these values with, and I installed Qt 5.12.11. Keep in mind that this data might only apply to that particular version of Qt.

When `qt.conf` is present, the patches have no effect.

When `qt.conf` is not present in the same directory as qmake,
1. `QT_HOST_PREFIX`, `QT_HOST_DATA`, `QT_HOST_BINS`, and `QT_HOST_LIBS` all become equal to the value at `qt_hpfxpath`, with or without the suffix `bin` or `lib`.
2. `QT_INSTALL_PREFIX`, `QT_INSTALL_ARCHDATA`, `QT_INSTALL_DATA`, `QT_INSTALL_DOCS`, `QT_INSTALL_HEADERS`, `QT_INSTALL_LIBS`, `QT_INSTALL_LIBEXECS`, `QT_INSTALL_BINS`, `QT_INSTALL_TESTS`, `QT_INSTALL_PLUGINS`, `QT_INSTALL_IMPORTS`, `QT_INSTALL_QML`, `QT_INSTALL_TRANSLATIONS`, `QT_INSTALL_EXAMPLES`, and `QT_INSTALL_DEMOS` all become equal to the value at `qt_epfxpath`, with or without a suffix like `bin` or `lib`.
3. `QT_INSTALL_PREFIX/dev`, `QT_INSTALL_ARCHDATA/dev`, `QT_INSTALL_DATA/dev`, `QT_INSTALL_DOCS/dev`, `QT_INSTALL_HEADERS/dev`, `QT_INSTALL_LIBS/dev`, `QT_INSTALL_LIBEXECS/dev`, `QT_INSTALL_BINS/dev`, `QT_INSTALL_TESTS/dev`, `QT_INSTALL_PLUGINS/dev`, `QT_INSTALL_IMPORTS/dev`, `QT_INSTALL_QML/dev`, `QT_INSTALL_TRANSLATIONS/dev`, `QT_INSTALL_EXAMPLES/dev`, `QT_INSTALL_DEMOS/dev` all become equal to the value at `qt_prfxpath`, with or without a suffix like `bin` or `lib`.

Given this information, I think it's important that #258 is merged, since it patches all three values. Patching `qt_prfxpath` alone appears to be insufficient.

For testing that these values have been patched, I think that either of these approaches would work:
1. Hide `qt.conf` and use `qmake -query`:
    * `qmake -query QT_INSTALL_PREFIX/dev` (for `qt_prfxpath`)
    * `qmake -query QT_INSTALL_PREFIX` (for `qt_epfxpath`)
    * `qmake -query QT_HOST_PREFIX` (for `qt_hpfxpath`)
2. Directly read the `qmake` binary, using a tool like `strings`:
    * `strings qmake | grep qt_prfxpath | cut -d '=' -f 2` (for `qt_prfxpath`)
    * `strings qmake | grep qt_epfxpath | cut -d '=' -f 2` (for `qt_epfxpath`)
    * `strings qmake | grep qt_hpfxpath | cut -d '=' -f 2` (for `qt_hpfxpath`)

Personally, I prefer the second approach, because of how direct it is, but I don't mind using `qmake -query` if that's your prefence.

What I'd really prefer is a test that causes a build failure when the values aren't patched properly, but I'm having trouble getting that to work. On a previous run, when I had `qt.conf` hidden from `qmake`, `qmake` was unable to locate the `mkspecs` directory,  which caused the build to fail.

**Comment by miurahr** [comment]
Now #258 merged. @ddalcino ,is #259 ready?

**Comment by ddalcino** [comment]
Not just yet. I'm going to squash some commits so the git history makes more sense first.

**Comment by ddalcino** [comment]
Ok, #259 is ready now.

**Comment by ddalcino** [comment]
@tsteven4, is this issue fixed now that #258 is merged?

**Comment by tsteven4** [comment]
This issue is fixed.  I tested f865f5f469e1d78d17ab32a08019403e7c68cb8d with both long and short paths in my application successfully.

**Comment by miurahr** [comment]
Thanks @tsteven4 and @ddalcino for hard work!
There is v1.2.0b3 released and will be genral v1.2.0.
